locale.c: Don't use LC_ALL unless platform has it

khwilliamson opened the pull request:
There was one function where LC_ALL could be used even if not available on the platform.

bram-perl commented on locale.c:
```diff
@@ -6531,21 +6531,27 @@ Perl_thread_locale_init()
-            setlocale(LC_ALL, NULL)));
-
+                           "new thread, initial locale is %s;"
+                           " calling setlocale(LC_ALL, \"C\"\n",
```
the message in the debug log is missing a closing `)`
i.e. it's logging:
```
new thread, initial locale is ........; calling setlocale(LC_ALL, "C"
```

khwilliamson replied:
fixed

bram-perl commented on locale.c:
```diff
@@ -6531,21 +6531,27 @@ Perl_thread_locale_init()
 
-#  else
+#  endif
+#  if defined(LC_ALL)
```
This (also) alters the behavior on Windows, I'm assuming that's intended but in that case it would be good to explicitly mention it in the commit message.

Before this commit:
- Win32 it called `_configthreadlocale(_ENABLE_PER_THREAD_LOCALE);`
- on everything else: it called `Perl_setlocale(LC_ALL, "C")`

Now the `Perl_setlocale(LC_ALL, "C")` (or the variant with the for loop) is also called on Windows.

khwilliamson replied:
That was a bug in Windows, but I  overlooked that I was fixing it.  Anyway, two commits are now done, so the messages are explicit

khwilliamson committed:
Win32 should start new threads in C locale

Prior to this commit, wrong cpp directives did not guarantee this.

khwilliamson committed:
locale.c: Use LC_ALL only if available on platform

These are the final (unless I missed something) cases where LC_ALL could
be referred to even if undefined on the system.
